bedrock/signals_and_queries/workflows.py

```diff
@@ -0,0 +1,109 @@
+import asyncio
+from collections import deque
+from datetime import timedelta
+from typing import Deque, List, Optional, Tuple
+
+from temporalio import workflow
+
+with workflow.unsafe.imports_passed_through():
+    from bedrock.shared.activities import BedrockActivities
+
+
+@workflow.defn
+class SignalQueryBedrockWorkflow:
+    def __init__(self) -> None:
+        # List to store prompt history
+        self.conversation_history: List[Tuple[str, str]] = []
+        self.prompt_queue: Deque[str] = deque()
+        self.conversation_summary = ""
+        self.chat_timeout: bool = False
+
+    @workflow.run
+    async def run(self, inactivity_timeout_minutes: int) -> str:
+        while True:
+            workflow.logger.info(
+                "Waiting for prompts... or closing chat after "
+                + f"{inactivity_timeout_minutes} minute(s)"
+            )
+
+            # Wait for a chat message (signal) or timeout
+            try:
+                await workflow.wait_condition(
+                    lambda: bool(self.prompt_queue),
+                    timeout=timedelta(minutes=inactivity_timeout_minutes),
+                )
+            # If timeout was reached
+            except asyncio.TimeoutError:
+                self.chat_timeout = True
+                workflow.logger.info("Chat closed due to inactivity")
+                # End the workflow
+                break
+
+            # Fetch next user prompt and add to conversation history
+            prompt = self.prompt_queue.popleft()
```

Comment:
You should process the whole queue here instead of just re-running the loop for each item. Usually re-running a loop would be no big deal, but this starts a timer each time for the timeout. So if you have 100 signals that come in at the same time, this code will start 100 timers as opposed to probably just 1. Arguably we should not even start timers if the lambda evaluates to true immediately, but we do.